### ISSUE TITLE
Blazor - Model Binding InputText and String

### ISSUE THREAD

**Issue by casper7526** [dotnet/aspnetcore#40785]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Public string? X
 
Bind the string to an InputText, post the form. X is null.
Type something into the field than delete the text, post the string. X is now an empty string.

Work around was just adding a check on my setter but the fact that an empty text box can represent both null and empty string is not clear.

### Expected Behavior

String should always be null when blank except when explicitly told otherwise.

### .NET Version

6

**Comment by casper7526** [comment]
The way I found this was by putting a minlength annotation on my string but not required. So it passes validation unless the string is modified.

**Comment by rafikiassumani-msft** [comment]
@casper7526 Can you provide a link to a GitHub repo to reproduce this ?

**Comment by casper7526** [comment]
https://github.com/casper7526/BlazorTextInputBug

Add Text to the InputText box and cause a Blur --- The value is now NOT NULL
Delete All Text from the InputText box and cause a Blur -- The value remains NOT NULL -- Empty.String

The problem is caused by the 2 following lines of code inside of 
Components/Web/src/Forms/InputBase.cs

```
 public override Task SetParametersAsync(ParameterView parameters)
......
_nullableUnderlyingType = Nullable.GetUnderlyingType(typeof(TValue));

This sets _nullableUnderlyingType to NULL when ran on a string? (reference type)
```

```
 protected string? CurrentValueAsString
    {
        get => FormatValueAsString(CurrentValue);
        set
        {
            _parsingValidationMessages?.Clear();

            bool parsingFailed;

            if (_nullableUnderlyingType != null && string.IsNullOrEmpty(value))
```

The first IF-CHECK could be triggered in this scenario and cause CurrentValue to be null.

I believe simply by changing the following the bug will be fixed and working as intended.

- if (_nullableUnderlyingType != null && string.IsNullOrEmpty(value))
+ if ((_nullableUnderlyingType != null || TValue is string) && string.IsNullOrEmpty(value))

**Comment by casper7526** [comment]
I'm not sure without lots of playing around, but this should ONLY happen with a string? (nullable)
If the property was to be a string, than possibly an empty string would be the expected behavior.

**Comment by javiercn** [comment]
I suspect this is expected as it only reflects the value from the underlying input element.

The initial value for the binding is null, when you type something, the value from the textbox is updated, and the binding is updated to reflect that. When you delete the text, the value (in JS) is the empty string, and the binding is updated accordingly.

For other data types doesn't behave that way because I believe the binding filters values that it can't parse and won't set the associated value. That doesn't happen with InputText because any value is always a valid value.

**Comment by casper7526** [comment]
Should this be marked as resolved?

**Comment by javiercn** [comment]
@casper7526 I think the system is behaving in the way it is expected to do so, it's not about having the same behavior other types expose, its about accepting valid or invalid values. For other types, `""` is not a valid value, but for string is.

**Comment by casper7526** [comment]
_nullableUnderlyingType != null && string.IsNullOrEmpty(value))

I think the line of code above implies that the intent was to make the value remain null.

In other model binding designs the value would return null as well, .net core mvc, framework, etc.

How would the possibility of a string returning null OR "" (at no choice to the developer) be considered "ok"

**Comment by casper7526** [comment]
TLDR - it does not behave the same way as other parts of .net binding and adds ambiguity to developers.

**Comment by mika76** [comment]
Not sure myself, but have you tried the [convertemptystringtonull](https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.dataannotations.displayformatattribute.convertemptystringtonull?view=net-6.0) attribute?

**Comment by casper7526** [comment]
No luck with that attribute.
